Add quickstart for minimal MCP Server using .NET and publish to NuGet (#47007)
dotnet/docs: docs/ai/quickstarts/build-mcp-server-publish-nuget.md

      "packages": [
        {
          "registry_name": "nuget",
          "name": "<your package ID here>",
          "version": "0.1.0-beta",
          "package_arguments": [],
          "environment_variables": [
            {
              "description": "The maximum number to return from the random number generator",
              "name": "MAX_RANDOM_NUMBER"
            }
          ]
        }
      ],
      "repository": {
        "url": "https://github.com/<your GitHub username here>/<your repo name>",
        "source": "github"
      },
      "version_detail": {
        "version": "0.1.0-beta"
      }
    }
    ```

    The `server.json` file schema is defined by the [MCP Registry project](https://github.com/modelcontextprotocol/registry) and is used by NuGet.org to generate VS Code MCP configuration.

1. Pack the project:

    ```bash
    dotnet pack -c Release
    ```

1. Publish the package to NuGet:

    ```bash
    dotnet nuget push bin/Release/*.nupkg --api-key <your-api-key> --source https://api.nuget.org/v3/index.json  
    ```

For more information, see [Publish a package](/nuget/nuget-org/publish-a-package).

## Discover MCP servers on NuGet.org

> 1. Search for your MCP Server package on [NuGet.org](https://www.nuget.org/packages?packagetype=mcpserver).

Reviewer: TODO: add screenshots from NuGet.org

---
title: Quickstart - Create a minimal MCP Server using .NET and publish to NuGet
description: Learn to create and connect to a minimal MCP server using .NET and publish it to NuGet.
ms.date: 06/30/2025
ms.topic: quickstart
ms.custom: devx-track-dotnet, devx-track-dotnet-ai
author: jondouglas
ms.author: jondouglas
---

> # Create, connect and publish a minimal MCP server using .NET to NuGet

Reviewer: Should we have a warning up here about this being a preview experience? Notably a preview C# MCP SDK (API shape may change) and needing preview .NET 10 SDK for now.

Author: i dont think we need to given how far along the release is already

In this quickstart, you create a minimal Model Context Protocol (MCP) server using the [C# SDK for MCP](https://github.com/modelcontextprotocol/csharp-sdk), connect to it using GitHub Copilot, and publish it to NuGet. MCP servers are services that expose capabilities to clients through the Model Context Protocol (MCP).

## Prerequisites

> - [.NET 10.0 SDK or higher](https://dotnet.microsoft.com/download)

Reviewer: I think we should mentioned .NET 10 Preview 6 here for now. If they have an earlier preview it won't work. We can clean i up later if needed.

Author: Ill change to the download/dotnet page as it actually has 10.0 listed, but this should be fine to not have to mention preview as it means we wont have to change the doc later.

- [Visual Studio Code](https://code.visualstudio.com/)
- [GitHub Copilot extension](https://marketplace.visualstudio.com/items?itemName=GitHub.copilot) for Visual Studio Code
- [NuGet account](https://www.nuget.org/account/login)

## Create the project

1. In a terminal window, install the `dotnet new mcpserver` template:

   ```bash

>    dotnet new install Microsoft.Extensions.AI.Templates::9.6.0-preview.2.25310.2

Reviewer: TODO: update to the correct package version

>    dotnet new install Microsoft.Extensions.AI.Templates::9.6.0-preview.2.25310.2

Reviewer 2: To reduce runtime dependencies and make nice RID-specific tools, you need to:

* set the `RuntimeIdentifiers` property with the RIDs of all the tools you want to make
* for p6, set `PublishSelfContained` property to `true`
  * if you want AOT, also set `PublishAot` to true - this requires [some orchestration](https://github.com/baronfel/multi-rid-tool/blob/main/.github/workflows/build-aot-packages.yml#L12-L97) though to make RID-specific tool packages
* for p7 and beyond you can be 
  * RID-specific but not self-contained if you want (`PublishSelfContained=false`)
  * Also add trimming (`PublishTrimmed=true`)

And then call `dotnet pack` on your tool project. If you're not AOTing, you'll end up making a bunch of packages that you upload to a NuGet feed. If you are AOTing, you need to make [RID-specific packages](https://github.com/baronfel/multi-rid-tool/blob/main/.github/workflows/build-aot-packages.yml#L13-L50) and a [top-level manifest package](https://github.com/baronfel/multi-rid-tool/blob/main/.github/workflows/build-aot-packages.yml#L52-L72) and push these up to a NuGet feed.

From there a user on 10p6 and upwards will be able to run `dotnet tool exec <toolname>` or `dnx <toolname>` and everything will just work.

>    dotnet new install Microsoft.Extensions.AI.Templates::9.6.0-preview.2.25310.2

Reviewer: This will be done in a future update to the template (and perhaps quick-start guide). Tracked here: https://github.com/dotnet/extensions/issues/6560

   ```

2. Create a new MCP server app with the `dotnet new mcpserver` command:

   ```bash
   dotnet new mcpserver -n MinimalMcpServer
   ```

3. Navigate to the `MinimalMcpServer` directory:

   ```bash
   cd MinimalMcpServer
   ```

4. Build the project:

   ```bash
   dotnet build
   ```

## Configure the MCP server in Visual Studio Code

Configure GitHub Copilot for Visual Studio Code to use your custom MCP server:

1. If you haven't already, open your project folder in Visual Studio Code.
2. Create a `.vscode` folder at the root of your project.
3. Add an `mcp.json` file in the `.vscode` folder with the following content:

   ```json
    {

>     "servers": {

Reviewer: Indentation looks a little weird on this line. "servers" should be at a deeper level than the first curly

          "MinimalMcpServer": {
            "type": "stdio",
            "command": "dotnet",
            "args": [
              "run",
              "--project",

>               "<path-to-your-csproj>",

Reviewer: I tried it. You can remove the `--project` and `<path-to-your-csproj>` args when using workspace settings file

>               "<path-to-your-csproj>",

Reviewer: Fixed.

2. View the MCP Server details and copy the json.

3. In your `mcp.json` file in the `.vscode` folder, add the following:

    ```json
    {   
      "servers": {
          "MinimalMcpServer": {
              "type": "stdio",
              "command": "dotnet",
              "args": [ "tool",
              "exec",
              "<your-package-id>",
              "--version",
              "0.0.1-beta",
              "--yes",
              "--",
              "start-mcp"

              ],
              "env": {
                  "MAX_RANDOM_NUMBER": 100
              }
          }
      }
    }
    ```

4. Save the file.

5. In GitHub Copilot, select the **Select tools** icon to verify your **MinimalMcpServer** is available with the tools listed.

6. Enter a prompt to run the **GetRandomNumber** tool:

    ```console
    Give me a random number between 1 and 100.
    ```

7. Verify that the server responds with the echoed message:

    ```output
    Your random number is 9.
    ```

## Related content

[Build a minimal MCP server](build-mcp-server.md)
[Get started with .NET AI and the Model Context Protocol](../get-started-mcp.md)

> [Use agent mode in VS Code](https://code.visualstudio.com/docs/copilot/chat/chat-agent-mode)

Author: https://learn.microsoft.com/visualstudio/ide/mcp-servers or https://learn.microsoft.com/visualstudio/ide/copilot-agent-mode ? 

@albarry4  may have thoughts too

---
title: Quickstart - Create a minimal MCP Server using .NET and publish to NuGet
description: Learn to create and connect to a minimal MCP server using .NET and publish it to NuGet.
ms.date: 06/30/2025
ms.topic: quickstart
ms.custom: devx-track-dotnet, devx-track-dotnet-ai
author: jondouglas
ms.author: jondouglas
---

> # Create, connect and publish a minimal MCP server using .NET to NuGet

Reviewer 3: ```suggestion
# Create, connect, and publish a minimal MCP server to NuGet
```

In this quickstart, you create a minimal Model Context Protocol (MCP) server using the [C# SDK for MCP](https://github.com/modelcontextprotocol/csharp-sdk), connect to it using GitHub Copilot, and publish it to NuGet. MCP servers are services that expose capabilities to clients through the Model Context Protocol (MCP).

## Prerequisites

> - [.NET 10.0 SDK or higher](https://dotnet.microsoft.com/download/dotnet)

Reviewer: ```suggestion
- [.NET 10.0 SDK or higher](https://dotnet.microsoft.com/download/dotnet) (preview 6 or newer)
```

              "MAX_RANDOM_NUMBER": 100
            }
          }
        }
    }
   ```

4. Save the file.

## Test the MCP server

1. Open GitHub Copilot in Visual Studio Code and switch to agent mode.

2. Select the **Select tools** icon to verify your **MinimalMcpServer** is available with both tools listed.

    :::image type="content" source="../media/mcp/available-tools-nuget.png" alt-text="A screenshot showing the available MCP tools.":::

3. Enter a prompt to run the **GetRandomNumber** tool:

    ```console
    Give me a random number between 1 and 100.
    ```

4. GitHub Copilot requests permission to run the **GetRandomNumber** tool for your prompt. Select **Continue** or use the arrow to select a more specific behavior:

    - **Current session** always runs the operation in the current GitHub Copilot Agent Mode session.
    - **Current workspace** always runs the command for the current Visual Studio Code workspace.
    - **Always allow** sets the operation to always run for any GitHub Copilot Agent Mode session or any Visual Studio Code workspace.

5. Verify that the server responds with the echoed message:

    ```output
    Your random number is 9.
    ```

> ## Pack and Publish to NuGet

Reviewer 3: ```suggestion
## Pack and publish to NuGet
```

1. Update the `<PackageId>` in the `.csproj` file to be unique on NuGet.org.

2. Update your `.mcp/server.json` file with your respective MCP configuration details.

    ```json
    {
      "description": "<your description here>",
      "name": "io.github.<your GitHub username here>/<your repo name>",
      "packages": [
        {
          "registry_name": "nuget",
          "name": "<your package ID here>",
          "version": "0.1.0-beta",
          "package_arguments": [

>             {

Author: protip use the suggestions button CTRL+G so its easy to commit! :)

>             {

Reviewer: I am not sure how to remove multiple lines with that

>             {

Author: fixed

              "description": "Start the MCP server tool in server mode",
              "is_required": true,
              "format": "string",
              "value": "start-mcp",
              "default": "start-mcp",
              "type": "positional",
              "value_hint": "start-mcp"
            }
          ],
          "environment_variables": [
            {
              "description": "The maximum number to return from the random number generator",
              "name": "MAX_RANDOM_NUMBER"
            }
          ]
        }
      ],
      "repository": {
        "url": "https://github.com/<your GitHub username here>/<your repo name>",
        "source": "github"
      },
      "version_detail": {
        "version": "0.1.0-beta"
      }
    }
    ```

3. Pack the project:

    ```bash
    dotnet pack -c Release
    ```

4. Publish the package to NuGet:

    ```bash

>     dotnet nuget push <path-to-nupkg-file> --source nuget.org --api-key <your-api-key>

Reviewer: ```suggestion
    dotnet nuget push bin/Release/*.nupkg  --api-key <your-api-key> --source https://api.nuget.org/v3/index.json
```

    ```

> See [Publish a package](https://learn.microsoft.com/nuget/nuget-org/publish-a-package) for more information.

Reviewer 3: ```suggestion
For more information, see [Publish a package](/nuget/nuget-org/publish-a-package).
```

## Discover MCP servers on NuGet.org

1. Browse for your MCP Server package on [NuGet.org](https://www.nuget.org/packages?q=&includeComputedFrameworks=true&packagetype=mcpserver&prerel=true&sortby=relevance).

> 2. View the MCP Server details and copy the json.

Reviewer 3: ```suggestion
2. View the MCP Server details and copy the JSON.
```

3. In your `mcp.json` file in the `.vscode` folder, add the following:

    ```json
    {   
      "servers": {
          "MinimalMcpServer": {
              "type": "stdio",
              "command": "dotnet",
              "args": [ "tool",
              "exec",
              "<your-package-id>",
              "--version",
              "0.0.1-beta",
              "--yes",
              "--",
              "start-mcp"

              ],
              "env": {
                  "MAX_RANDOM_NUMBER": 100
              }
          }
      }
    }
    ```

4. Save the file.

5. In GitHub Copilot, select the **Select tools** icon to verify your **MinimalMcpServer** is available with the tools listed.

6. Enter a prompt to run the **GetRandomNumber** tool:

    ```console
    Give me a random number between 1 and 100.
    ```

7. Verify that the server responds with the echoed message:

    ```output
    Your random number is 9.
    ```

## Related content

> [Build a minimal MCP server](build-mcp-server.md)

Reviewer 3: This should probably be a bulleted list, otherwise it's all put on one line (you can see that [here](https://learn.microsoft.com/en-us/dotnet/ai/quickstarts/build-mcp-server#related-content) where it should be fixed too @alexwolfmsft).

Author: added!

[Get started with .NET AI and the Model Context Protocol](../get-started-mcp.md)
[Use agent mode in VS Code](https://code.visualstudio.com/docs/copilot/chat/chat-agent-mode)

> [Use agent mode in Visual Studio](https://learn.microsoft.com/visualstudio/ide/copilot-agent-mode)
> [Model Context Protocol .NET Samples](https://github.com/microsoft/mcp-dotnet-samples)
> [What's New in .NET 10](https://learn.microsoft.com/dotnet/whats-new/dotnet-10)
> [Find and evaluate NuGet packages for your project](https://learn.microsoft.com/en-us/nuget/consume-packages/finding-and-choosing-packages)
> [Publish a package](https://learn.microsoft.com/nuget/nuget-org/publish-a-package)

Reviewer 3: ```suggestion
[Use agent mode in Visual Studio](/visualstudio/ide/copilot-agent-mode)
[Model Context Protocol .NET Samples](https://github.com/microsoft/mcp-dotnet-samples)
[What's New in .NET 10](../../core/whats-new/dotnet-10/overview.md)
[Find and evaluate NuGet packages for your project](/nuget/consume-packages/finding-and-choosing-packages)
[Publish a package](/nuget/nuget-org/publish-a-package)
```
